teps/0121-refine-retries-for-taskruns-and-customruns.md

```diff
@@ -113,43 +119,60 @@ Several observations regarding to the feature table above:
+### Timeout per Retry
 
-### Option 1: Implement `retries` for Standalone TaskRun
+Task-level Timeout (`TaskRunSpec.Timeout` and `RunSpec.Timeout`) is set for each `Retry` attempt.
 
-- Stop relying on `len(retriesStatus)` to determine whether a TaskRun or CustomRun finishes, use `ConditionSucceeded` & `ConditionFalse` & Reason=="TimedOut" instead.
-- `Retries` and `Timeout` are passed from `PipelineTask` to `TaskRunSpec` and `CustomRunSpec`.
+### Retries in TaskRuns and CustomRuns
 
-Three sub-options about the way to implement `retriesStatus`:
+Add a `Retries` field to `TaskRunSpec` and `CustomRunSpec`. Note that `Retries` already exists in `RunSpec`. The `TaskRun` and `CustomRun` controllers handle their own `Retries`. The `PipelineTask.Retries` value, which is specified at `Pipeline` authoring time, is passed to the `TaskRunSpec.Retries` and `CustomRunSpec.Retries` during the execution of a `PipelineRun`. 
 
-- 1.a: Update `retriesStatus` for each retry attempt for `TaskRun`, keep `retriesStatus` for `CustomRun`
-  - No API change
-  - Need to implement a strategy for clients to get the previous pod and read its logs.
+The `PipelineRun` controller stops relying on `len(retriesStatus)` to determine whether a `TaskRun` or `CustomRun` is done executing. Instead, uses `ConditionSucceeded` as the only way to decide if the `TaskRun` or `CustomRun` has completed execution.
 
-- 1.b: Update `retriesStatus` for each retry attempt for `TaskRun`, deprecate `retriesStatus` for `CustomRun`
-  - No implementation restrictions of `retriesStatus` for `CustomRun`
-  - Need to implement a strategy for clients to get the previous pod and read its logs.
+### Conditions.Succeeded
 
-- 1.c: Deprecate `retriesStatus` for both `TaskRun` and `CustomRun`, create a new `TaskRun` for each retry attempt, add a new field `RetryAttempts` in `TaskRunStatusFields` to record names of all retry attempts.
-  - Easier to retrieve logs from retried TaskRuns.
-  - See [Appendix - I](#i-some-implementation-details-about-option-1c) for more implementation details.
+The `TaskRun` and `CustomRun` controllers MUST set `Conditions.Succeeded` to `False` only upon eventual failure of the `TaskRun` or `CustomRun` when all the Retries have been exhausted. We will implement this behavior for `TaskRuns` and clearly document this requirement for `CustomRuns`.
 
-**Benefits:**
+This is a change to meaning of `Conditions.Succeeded` for `TaskRuns` so this change is a blocker for V1 release.
+
+### RetriesStatus
+
+Keep `RetriesStatus` for both `CustomRun` and `TaskRun` to hold information that's useful for CLI and Dashboard. Note that this field is not required. Custom Task implementers have the freedom to implement the `Retries` as what they want.
+
+## Future Work
```

Comment:
The current way retries are implemented is not so nice because we delete details about intermediate retries.
A solution to that would be to use separate `TaskRuns`, one per retry. 
With the proposed solution (i.e. retries managed by the taskrun/customrun controller), that is not an option, we must use a single `TaskRun` to manage all retries, and thus we need to make sure that we keep enough information in the status to let users know the current status, how many retries where executed until now, the result and logs of each retry.

I think whether we use `RetriesStatus` or `Conditions.RetrySucceeded` or something else is something else to fix the issue is something we can define on the related TEP. I think it's fine to have a proposal in here but it shouldn't be considered as a decision for future work.

Reply:
Makes sense, thanks Andrea! Moving it as an alternative.